apps/api/src/http/internal/rest/webhooks/paypal/sale.rs

```diff
@@ -0,0 +1,279 @@
+		.find_one(
+			filter::filter! {
+				SubscriptionProduct {
+					#[query(flatten)]
+					variants: SubscriptionProductVariant {
+						paypal_id: Some(&paypal_sub.plan_id),
+					}
+				}
+			},
+			None,
+		)
+		.await?
+	else {
+		// no product found
+		return Ok(());
+	};
+
+	let stripe_product_id = product
+		.variants
+		.into_iter()
+		.find(|v| v.paypal_id.as_ref().is_some_and(|p| p == &paypal_sub.plan_id))
+		.unwrap()
+		.id;
+
+	let invoice = stripe::Invoice::create(
+		&global.stripe_client,
+		CreateInvoice {
+			customer: Some(customer_id.clone().into()),
+			auto_advance: Some(false),
+			description: Some("Legacy PayPal invoice. Real payments will be handled by PayPal."),
+			metadata: Some(std::iter::once(("PAYPAL_ID".to_string(), sale.id.clone())).collect()),
+			..Default::default()
+		},
+	)
+	.await
+	.map_err(|e| {
+		tracing::error!(error = %e, "failed to create invoice");
+		TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+	})?;
+
+	stripe::Invoice::finalize(
+		&global.stripe_client,
+		&invoice.id,
+		FinalizeInvoiceParams {
+			auto_advance: Some(false),
+		},
+	)
+	.await
+	.map_err(|e| {
+		tracing::error!(error = %e, "failed to finalize invoice");
+		TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+	})?;
+
+	let invoice = stripe::Invoice::void(&global.stripe_client, &invoice.id).await.map_err(|e| {
+		tracing::error!(error = %e, "failed to void invoice");
+		TransactionError::custom(ApiError::INTERNAL_SERVER_ERROR)
+	})?;
+
+	let invoice_id: InvoiceId = invoice.id.into();
+
+	let status = invoice.status.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?.into();
+
+	let created_at = chrono::DateTime::from_timestamp(invoice.created.unwrap_or_default(), 0)
+		.ok_or(TransactionError::custom(ApiError::BAD_REQUEST))?;
+
+	tx.insert_one(
+		Invoice {
+			id: invoice_id.clone(),
+			items: vec![stripe_product_id],
+			customer_id: customer_id.into(),
+			user_id: user.id,
+			paypal_payment_id: Some(sale.id.clone()),
+			status,
+			failed: false,
+			refunded: false,
+			disputed: None,
+			created_at,
+			updated_at: created_at,
+			search_updated_at: None,
+		},
+		None,
+	)
+	.await?;
+
+	if let Some(next_billing_time) = paypal_sub.billing_info.next_billing_time {
+		let subscription_id = SubscriptionId {
+			user_id: user.id,
+			product_id: product.id,
+		};
+
+		tx.insert_one(
+			SubscriptionPeriod {
+				id: SubscriptionPeriodId::new(),
+				subscription_id: subscription_id.clone(),
+				provider_id: Some(ProviderSubscriptionId::Paypal(provider_id)),
+				start: paypal_sub
+					.billing_info
+					.last_payment
+					.map(|p| p.time)
+					.unwrap_or_else(chrono::Utc::now),
+				end: next_billing_time,
+				is_trial: false,
+				created_by: SubscriptionPeriodCreatedBy::Invoice { invoice_id },
+				updated_at: chrono::Utc::now(),
+				search_updated_at: None,
+			},
+			None,
+		)
+		.await?;
+
+		sub_refresh_job::refresh_entitlements(&mut tx, &subscription_id, &product.benefits).await?;
```

Comment:
This should likely not be inside the transaction and should just be run after the tx.